cpp/oneapi/dal/algo/decision_tree/common.hpp

```diff
@@ -0,0 +1,193 @@
+/*******************************************************************************
+* Copyright 2021 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#pragma once
+
+#include "oneapi/dal/detail/common.hpp"
+#include "oneapi/dal/table/common.hpp"
+#include "oneapi/dal/util/common.hpp"
+
+namespace oneapi::dal::decision_tree {
+
+namespace task {
+namespace v1 {
+/// Tag-type that parameterizes entities used for solving
+/// :capterm:`classification problem <classification>`.
+struct classification {};
+
+/// Tag-type that parameterizes entities used for solving
+/// :capterm:`regression problem <regression>`.
+struct regression {};
+
+/// Alias tag-type for classification task.
+using by_default = classification;
+} // namespace v1
+
+using v1::classification;
+using v1::regression;
+using v1::by_default;
+
+} // namespace task
+
+namespace detail {
+namespace v1 {
+
+template <typename Task>
+class node_info_impl;
+
+template <typename Task>
+class split_node_info_impl;
+
+template <typename Task>
+class leaf_node_info_impl;
+
+template <typename Task>
+constexpr bool is_valid_task_v =
+    dal::detail::is_one_of_v<Task, task::classification, task::regression>;
+
+} // namespace v1
+
+using v1::node_info_impl;
+using v1::split_node_info_impl;
+using v1::leaf_node_info_impl;
+using v1::is_valid_task_v;
+} // namespace detail
+
+namespace v1 {
+
+/// Class containing base node info in decision tree
+template <typename Task = task::by_default>
+class node_info : public base {
+    static_assert(detail::is_valid_task_v<Task>);
+    friend dal::detail::pimpl_accessor;
+
+    using task_t = Task;
+
+public:
+    node_info();
+    node_info(const node_info<task_t>&);
+    node_info<task_t>& operator=(const node_info<task_t>&);
```

Comment:
@makart19, I suggest to implement variant of `pimpl` class that supports copy by value semantics, but not duplicate these operators in each class. Moreover, there are rvalue reference cases not covered here.

Reply:
migrated to simple pointer as discussed